src/ott/solvers/nn/losses.py

```diff
@@ -0,0 +1,85 @@
+# Copyright OTT-JAX
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import jax
+import jax.numpy as jnp
+from ott.geometry import pointcloud
+from ott.solvers.linear import sinkhorn
+from ott.problems.linear import linear_problem
+import ott.geometry.costs as costs
+from typing import Any, Mapping, Callable
+from types import MappingProxyType
+
+class MongeGap:
+    """
+    A class to define the Monge gap regularizer.
+    For a cost function :`math:`c` and an empirical reference :math:`\rho_x`
+    defined by samples :math:`x_i`, the (entropic) Monge gap of a vector field :math:`T` 
+    is defined as:
+        :math:`\mathcal{M}^c_{\rho_x, \epsilon} = \frac{1}{n} \sum_{i=1}^n c(x_i, T(x_i)) 
+        - W_\epsilon(\mu_x, T \# \rho_x)`.
+    
+    Args:
+    geometry_kwargs: Holds the kwargs to instanciate the geometry to compute the ``reg_ot_cost``. 
+                     Default cost function (if no cost function is passed in ``geometry_kwargs``)
+                     is the quadratic costs ``ott.geometry.costs.SqEuclidean()``.
+    sinkhorn_kwargs: Holds the kwargs to instanciate the Sinkhorn solver to compute the ``reg_ot_cost``.
+    """
+
+    def __init__(
+        self,
+        geometry_kwargs: Mapping[str, Any] = MappingProxyType({}),
+        sinkhorn_kwargs: Mapping[str, Any] = MappingProxyType({}),
+    ) -> None:
+        self.geometry_kwargs = geometry_kwargs
+        self.sinkhorn_kwargs = sinkhorn_kwargs
+
+    def __call__(
+        self, samples: jnp.ndarray, T: Callable[[jnp.ndarray], jnp.ndarray]
```

Comment:
I don't think this function should receive the function computing the transport, as it limits what can be passed (e.g., classes implementing `__call__`). I'd rather you pass the transported samples.

Reply:
I think it might be worth revisiting this choice. 

In fact, I think the notion of _Monge gap_ only has meaning when it is passed with callable (the Monge gap is defined as a function defined on maps in the paper), so I think the initial implementation was better conceptually (i.e. I would pass callable `T` and `source` points as arguments, just to instantiate `target` as being `T` run on samples. 

We can have a function with `source` and `target` points passed directly, but this should be called something different (e.g. `id_coupling_gap`).